src/dbnode/storage/index/compaction/plan.go

```diff
@@ -0,0 +1,250 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package compaction
+
+import (
+	"errors"
+	"fmt"
+	"sort"
+	"time"
+
+	"github.com/m3db/m3/src/dbnode/storage/index/segments"
+)
+
+var (
+	errMutableCompactionAgeNegative = errors.New("mutable compaction age must be positive")
+	errLevelsUndefined              = errors.New("compaction levels are undefined")
+	errMaxImmutableCompactionSize   = errors.New("max immutable compaction size must be positive")
+)
+
+var (
+	// DefaultLevels are the default Level(s) used for compaction options.
+	DefaultLevels = []Level{ // i.e. tiers for compaction [0, 262K), [262K, 524K), [524K, 1M), [1M, 2M), [2M, 8M)
+		Level{
+			MinSizeInclusive: 0,
+			MaxSizeExclusive: 1 << 18,
+		},
+		Level{
+			MinSizeInclusive: 1 << 18,
+			MaxSizeExclusive: 1 << 19,
+		},
+		Level{
+			MinSizeInclusive: 1 << 19,
+			MaxSizeExclusive: 1 << 20,
+		},
+		Level{
+			MinSizeInclusive: 1 << 20,
+			MaxSizeExclusive: 1 << 21,
+		},
+		Level{
+			MinSizeInclusive: 1 << 21,
+			MaxSizeExclusive: 1 << 23,
+		},
+	}
+
+	// DefaultOptions are the default compaction PlannerOptions.
+	DefaultOptions = PlannerOptions{
+		MaxImmutableCompactionSize: 1 << 22,                            // ~4M
+		MaxMutableSegmentSize:      1 << 16,                            // 64K
+		MutableCompactionAge:       15 * time.Second,                   // any mutable segment 15s or older is eligible for compactions
```

Comment:
Ah interesting.  Yeah end to end latency is the primary concern, also at very low volume it's annoying that you have to wait a fixed amount for the series to become visible.  Ideally we should be as fast as possible and only once we are performing substantially high workloads do we slow down.

Could we do something like if there's no mutable segment compactions happening then kick one off, but otherwise buffer?  So that we're doing it "as fast as possible" but should also avoid tons of tiny segments?

Reply:
I really like it. I've been mulling a way to make a feedback loop in the compaction process by modelling `f(cost of compaction, compaction budget, latency)` but the version you suggested is a great start. 

I'll make the following changes in the PR:
- Decompose identifying "compactable" segments from the creation of the plan 
- Add `func (s Segment) Compactable(opts PlannerOptions) bool` (w/ the same conditions as they are in `NewPlan` currently) 

We can land this PR after that. In #819, I'll make `indexBlock.triggerCompactions` consider both `segment.Compactible` and whether the there's an on-going compaction

Reply:
done